aten/src/ATen/native/cuda/group_norm_kernel.cu

```diff
@@ -0,0 +1,610 @@
+      }
+    }
+  }
+  g_shared[threadIdx.y][threadIdx.x] = dg_sum1;
+  g_shared[threadIdx.y + blockDim.y][threadIdx.x] = dg_sum2;
+  b_shared[threadIdx.y][threadIdx.x] = db_sum1;
+  b_shared[threadIdx.y + blockDim.y][threadIdx.x] = db_sum2;
+  __syncthreads();
+  T_ACC sum1 = g_shared[threadIdx.x][threadIdx.y];
+  T_ACC sum2 = b_shared[threadIdx.x][threadIdx.y];
+  sum1 = cuda_utils::WarpReduceSum<T_ACC>(sum1);
+  sum2 = cuda_utils::WarpReduceSum<T_ACC>(sum2);
+  if (threadIdx.x == 0) {
+    const int64_t c = blockIdx.x * blockDim.x + threadIdx.y;
+    if (c < C) {
+      if (dgamma != nullptr) {
+        dgamma[c] = sum1;
+      }
+      if (dbeta != nullptr) {
+        dbeta[c] = sum2;
+      }
+    }
+  }
+  sum1 = g_shared[threadIdx.x][threadIdx.y + blockDim.y];
+  sum2 = b_shared[threadIdx.x][threadIdx.y + blockDim.y];
+  sum1 = cuda_utils::WarpReduceSum<T_ACC>(sum1);
+  sum2 = cuda_utils::WarpReduceSum<T_ACC>(sum2);
+  if (threadIdx.x == 0) {
+    const int64_t c = blockIdx.x * blockDim.x + threadIdx.y + blockDim.y;
+    if (c < C) {
+      if (dgamma != nullptr) {
+        dgamma[c] = sum1;
+      }
+      if (dbeta != nullptr) {
+        dbeta[c] = sum2;
+      }
+    }
+  }
+}
+
+template <typename T>
+void GroupNormKernelImplInternal(
+    const Tensor& X,
+    const Tensor& gamma,
+    const Tensor& beta,
+    int64_t N,
+    int64_t C,
+    int64_t HxW,
+    int64_t group,
+    T eps,
+    Tensor* Y,
+    Tensor* mean,
+    Tensor* rstd) {
+  using T_ACC = acc_type<T, true>;
+  TORCH_CHECK(X.numel() == N * C * HxW);
+  TORCH_CHECK(!gamma.defined() || gamma.numel() == C);
+  TORCH_CHECK(!beta.defined() || beta.numel() == C);
+  if (N == 0) {
+    return;
+  }
+  const int64_t G = group;
+  const int64_t D = C / G;
+  const T* X_data = X.data_ptr<T>();
+  const T* gamma_data = gamma.defined() ? gamma.data_ptr<T>() : nullptr;
+  const T* beta_data = beta.defined() ? beta.data_ptr<T>() : nullptr;
+  T* Y_data = Y->data_ptr<T>();
+  T* mean_data = mean->data_ptr<T>();
+  T* rstd_data = rstd->data_ptr<T>();
+  const auto kAccType = X.scalar_type() == kHalf ? kFloat : X.scalar_type();
+  Tensor a = at::empty({N, C}, X.options().dtype(kAccType));
+  Tensor b = at::empty({N, C}, X.options().dtype(kAccType));
+  T_ACC* a_data = a.data_ptr<T_ACC>();
+  T_ACC* b_data = b.data_ptr<T_ACC>();
+  cudaStream_t cuda_stream = at::cuda::getCurrentCUDAStream();
+  RowwiseMomentsCUDAKernel<T>
+      <<<N * G, cuda_utils::kCUDABlockReduceNumThreads, 0, cuda_stream>>>(
+          D * HxW, eps, X_data, mean_data, rstd_data);
+  int64_t B = (N * C + kCUDANumThreads - 1) / kCUDANumThreads;
+  ComputeFusedParamsCUDAKernel<T><<<B, kCUDANumThreads, 0, cuda_stream>>>(
+      N, C, G, mean_data, rstd_data, gamma_data, beta_data, a_data, b_data);
+  if (HxW < kCUDANumThreads) {
+    B = (N * C * HxW + kCUDANumThreads - 1) / kCUDANumThreads;
+    GroupNormForwardSimpleCUDAKernel<T><<<B, kCUDANumThreads, 0, cuda_stream>>>(
```

Comment:
is it better for perf?

Reply:
Yes, since when the feature map becomes smaller by layers, some layers only has feature map whose size is 7 * 7 or 14 * 14, in that case, some of the threads are not working at all. In these cases, this will help.